README.md
![Build status](https://ci.appveyor.com/api/projects/status/9ftybkutu10h8yvl?svg=true)

# HPEPduPS
Powershell module as a wrapper for the REST Api in HPE G2 Metered and switched PDUs

The module currently includes four functions for working with the PDU API:
- Get-HPEPdu
    - Get information about the PDUs connected to the system
- Get-HPEPduLoadMeasurement
    - Get current load on the specified PDU
- Get-HPEPduOutlet
    - Get information about outlets on the specified segments
- Get-HPEPduOutletMeasurement
    - Get current load on the specified Outlet

A few examples on the usage:
```powershell
PS C:\> Get-Command -Module HPEPduPS

CommandType     Name                                               Version    Source
-----------     ----                                               -------    ------
Function        Get-HPEPdu                                         1.0.0      HPEPduPS
Function        Get-HPEPduLoadMeasurement                          1.0.0      HPEPduPS
Function        Get-HPEPduOutlet                                   1.0.0      HPEPduPS
Function        Get-HPEPduOutletMeasurement                        1.0.0      HPEPduPS
```

```powershell
PS C:\> Get-HPEPdu -System 1.1.1.1 -Username admin
Please specify password: *********

Id Path
-- ----
1  /redfish/v1/PowerDistribution/1
2  /redfish/v1/PowerDistribution/2
3  /redfish/v1/PowerDistribution/3
4  /redfish/v1/PowerDistribution/4
```

The functions support both username/password auth as well as a PS credential object
```powershell
PS C:\> Get-HPEPdu -System 1.1.1.1 -PduId 1 -Credential $cred


Core_location           :
@odata.context          : /redfish/v1/$metadata#PowerDistribution
BreakerRating           : 0
InputRating             : 16
Model                   : 230V, 16A, 11.0kVA, 50/60Hz
LoadsegmentMeasurement  : {@{@odata.id=/redfish/v1/PowerDistribution/1/PowerMeasurement/LoadsegmentMeasurement}}
Core_u_position         :
@odata.type             : #PowerDistribution.1.0.0.PowerDistribution
Panel_name              :
@odata.id               : /redfish/v1/PowerDistribution/1
OutletMeasurement       : {@{@odata.id=/redfish/v1/PowerDistribution/1/PowerMeasurement/Loadsegment/1/OutletMeasurement}, @{@odata.id=/redfish/
                          v1/PowerDistribution/1/PowerMeasurement/Loadsegment/2/OutletMeasurement}, @{@odata.id=/redfish/v1/PowerDistribution/1
                          /PowerMeasurement/Loadsegment/3/OutletMeasurement}, @{@odata.id=/redfish/v1/PowerDistribution/1/PowerMeasurement/Load
                          segment/4/OutletMeasurement}...}
Firmware_version        : 2.0.0.C
Serial                  : CNxxxx
Id                      : 1
PartNumber              : P9S20A
DeviceType              : PowerDistributionUnit
PowerDistributionNumber : 1
Boot_version            : 2.25
Hardware_version        : HPE
Voltage                 : 240
KVARating               : 11
Power_rating            : 11,0
OutletControl           : {@{@odata.id=/redfish/v1/PowerDistribution/1/PowerControl/Loadsegment/1/OutletControl}, @{@odata.id=/redfish/v1/Power
                          Distribution/1/PowerControl/Loadsegment/2/OutletControl}, @{@odata.id=/redfish/v1/PowerDistribution/1/PowerControl/Lo
                          adsegment/3/OutletControl}, @{@odata.id=/redfish/v1/PowerDistribution/1/PowerControl/Loadsegment/4/OutletControl}...}

```

All functions have help functionality included

For more information about the module check out my [blog post](https://www.rudimartinsen.com/2019/01/04/hpe-pdu-powershell-module/) about it

Please refer to this [repository](https://github.com/rumart/hpe-g2-pdu-api) for more information about the PDU Rest API.

[Changelog](https://github.com/rumart/HPEPduPS/blob/master/changelog.md)

changelog.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Added
- More Pester tests
- Support more functionality from API
- Output format

## [version 1.0.0] - 2019-01-04
### Changed
- Restructuring module, moving functions out of psm1 in to folders

## [version 0.4.1] - 2018-12-17
### Added
- Appveyor config and connection for CI

## [version 0.4.0] - 2018-12-17
### Added
- Pester test for valid Powershell and module import

## [version 0.3.1] - 2018-12-17
### Fixed
- Fixed bug in parameter name in Invoke-Restmethod call inside the Invoke-HpePduRequest function

## [version 0.3.0] - 2018-12-17
### Changed
- Setting system, PDU Id and Segment Id as a mandatory parameter

## [version 0.2.3] - 2018-12-17
### Changed
- Added check on username and credential parameter

## [version 0.2.2] - 2018-11-18
### Changed
- Updated changelog with Keep a changelog changes

## [version 0.2.1] - 2018-11-17
### Added
- Help text added to all functions

## [version 0.2.0] - 2018-11-17
### Added

- All functions support the use of a PSCredential object which you can create before you use the functions and then pass this object to each function call. Username/password will still work as in the previous version

## [version 0.1.0]

Initial creation of the module with the following functions available.

* Get-HPEPdu
* Get-HPEPduLoadMeasurement
* Get-HPEPduOutlet
* Get-HPEPduOutletMeasurement
* Invoke-HPEPduRequest
* Set-InsecureSSL

The two latter functions are internal/private to the module

With the functions you can perform actions like:
* List all PDUs connected to the Management system (Get-HPEPdu)
* Retrieve details about a specific PDU (Get-HPEPdu -PduId x)
* Retrieve the outlets of a specific segment on a specific PDU (Get-HPEPduOutlet -PduId x -SegmentId y)
* Retrieve the load of a PDU (Get-HPEPduLoadMeasurement -PduId x)
* Retrieve the measurements of the outlets on a segment (Get-HPEPduOutletMeasurement -PduId x -SegmentId y)

All functions require you to authenticate (there is no "Connect" function). The username and password needs to be passed to each function call

tests/Project.Tests.ps1
#This test expects the module to reside in a subfolder of the projectroot with the same name as the projectroot
$projectRoot = Resolve-Path "$PSScriptRoot\.."
$moduleName = Split-Path $projectRoot -Leaf
$moduleRoot = Split-Path (Resolve-Path "$projectRoot\$moduleName\*.psm1")

Describe "General project validation: $moduleName" {

    $scripts = Get-ChildItem $projectRoot -Include *.ps1,*.psm1,*.psd1 -Recurse

    # TestCases are splatted to the script so we need hashtables
    $testCase = $scripts | Foreach-Object{@{file=$_}}         
    It "Script <file> should be valid powershell" -TestCases $testCase {
        param($file)

        $file.fullname | Should Exist

        $contents = Get-Content -Path $file.fullname -ErrorAction Stop
        $errors = $null
        $null = [System.Management.Automation.PSParser]::Tokenize($contents, [ref]$errors)
        $errors.Count | Should Be 0
    }

    It "Module '$moduleName' can import cleanly" {
        {Import-Module (Join-Path $moduleRoot "$moduleName.psd1") -force } | Should Not Throw
    }
}

HPEPduPS/public/Get-HPEPduOutlet.ps1

function Get-HPEPduOutlet {
    <#
        .SYNOPSIS
            Get information about HPE Pdu Outlets
        .DESCRIPTION
            This function works with HPE G2 Metered and Switched Power distribution units.
            Retrieve information about the Outlets of the Pdu specified.
        .NOTES
            Info
            Author : Rudi Martinsen / Intility AS
            Date : 17/11-2018
            Version : 1.0.0
            Revised : 04/01-2019
            Changelog:
            1.0.0 -- Bumped version number
            0.3.0 -- Setting params as mandatory
            0.2.2 -- Added check on credentials
            0.2.1 -- Added help text
            0.2.0 -- Fixed support for credential object
        .LINK
            https://github.com/rumart/hpe-g2-pdu-api
        .LINK
            https://www.rudimartinsen.com/2018/11/19/exploring-the-hpe-g2-pdu-rest-api/
        .LINK
            https://www.rudimartinsen.com/2019/01/04/hpe-pdu-powershell-module/
        .EXAMPLE
            PS C:\> Get-HPEPduOutlet -System 10.10.10.10 -PDUId 1 -SegmentId 1 -Credential $credential

            OutletNumber : 2
            StartupState : on
            OutletName   : OUTLET 2
            OnDelay      : 0
            OffDelay     : 0
            RebootDelay  : 5
            OutletStatus : on

            OutletNumber : 3
            StartupState : on
            OutletName   : OUTLET 3
            OnDelay      : 0
            OffDelay     : 0
            RebootDelay  : 5
            OutletStatus : on

            OutletNumber : 4
            StartupState : on
            OutletName   : OUTLET 4
            OnDelay      : 0
            OffDelay     : 0
            RebootDelay  : 5
            OutletStatus : on

            OutletNumber : 5
            StartupState : on
            OutletName   : OUTLET 5
            OnDelay      : 0
            OffDelay     : 0
            RebootDelay  : 5
            OutletStatus : on

    #>
    [cmdletbinding()]
    param(
        [Parameter(Mandatory=$true)]
        $System,
        [Parameter(Mandatory=$true)]
        [int]
        $PDUId,
        [Parameter(Mandatory=$true)]
        [int]
        $SegmentId,
        [Parameter(Mandatory=$false,ParameterSetName="username")]
        $Username,
        [Parameter(Mandatory=$false,ParameterSetName="username")]
        [securestring]
        $Password,
        [Parameter(Mandatory=$false,ParameterSetName="credential")]
        $Credential,
        [switch]
        $IgnoreSSL
    )
    
    if($Username){
        if(!$Password -or $Password -eq ""){
            $Password = (Read-Host -Prompt "Please specify password" -AsSecureString)
        }
        $Credential = New-Object System.Management.Automation.PSCredential($Username,$Password)
    }
    elseif($Credential){
    }
    else{
        Write-Error "Username or Credential object not specified. Please specify valid credentials for connecting to the PDU"
    }

    $Path = "/redfish/v1/PowerDistribution/$PduId/PowerControl/Loadsegment/$SegmentId/OutletControl"

    $response = Invoke-HPEPduRequest -System $System -Resource $Path -Credential $Credential -IgnoreSSL

    if($response.'@odata.type' -eq '#OutletControl.1.0.0.OutletControl'){
        $response.Outlets
    }
    
}

HPEPduPS/private/Invoke-HPEPduRequest.ps1
function Invoke-HPEPduRequest {
    <#
        .SYNOPSIS
            Function for sending requests to the HPE PDU Api
        .DESCRIPTION
            This function is ment as an internal function to the other functions
            in the powershell module
        .NOTES
            Info
            Author : Rudi Martinsen / Intility AS
            Date : 17/11-2018
            Version : 1.0.0
            Revised : 04/01-2019
            Changelog:
            1.0.0 -- Bumped version number
            0.2.4 -- Fixed wrong param name in API request
            0.2.3 -- Outputting error message in verbose
            0.2.2 -- Putting response in variable
            0.2.1 -- Added help text
            0.2.0 -- Fixed support for credential object
        .LINK
            https://github.com/rumart/hpe-g2-pdu
        .LINK
            https://www.rudimartinsen.com/2018/11/19/exploring-the-hpe-g2-pdu-rest-api/
        .LINK
            https://www.rudimartinsen.com/2019/01/04/hpe-pdu-powershell-module/
    #>
    [cmdletbinding()]
    param(
        $System,
        $Resource,
        $Username,
        $Password,
        $Credential,
        [switch]
        $IgnoreSSL
    )

    if($Credential){
        $auth = [System.Convert]::ToBase64String([System.Text.Encoding]::UTF8.GetBytes($Credential.UserName+':'+$Credential.GetNetworkCredential().Password))
    }
    else{
        $unsecPass = [Runtime.InteropServices.Marshal]::PtrToStringAuto([Runtime.InteropServices.Marshal]::SecureStringToBSTR($Password))
        $auth = [System.Convert]::ToBase64String([System.Text.Encoding]::UTF8.GetBytes($UserName+':'+$UnsecPass))
    }

    $header = @{
      'Authorization' = "Basic $auth"
    }
    $uri = "https://$system" + $Resource

    if($IgnoreSSL){
        Set-InsecureSSL
    }

    $response = Invoke-RestMethod -Method Get -Uri $uri -Headers $header -ErrorVariable apiErr
    if($apiErr){
        Write-Verbose $apiErr.InnerException
    }
    $response
}

HPEPduPS/public/Get-HPEPduLoadMeasurement.ps1

function Get-HPEPduLoadMeasurement {
    <#
        .SYNOPSIS
            Get load measurements of HPE Pdu segments
        .DESCRIPTION
            This function works with HPE G2 Metered and Switched Power distribution units.
            Retrieves load measurements of the segments of the Pdu specified.
        .NOTES
            Info
            Author : Rudi Martinsen / Intility AS
            Date : 17/11-2018
            Version : 1.0.0
            Revised : 04/01-2019
            Changelog:
            1.0.0 -- Bumped version number
            0.3.0 -- Setting pduid as a mandatory variable
            0.2.2 -- Added check on credentials
            0.2.1 -- Added help text
            0.2.0 -- Fixed support for credential object
        .LINK
            https://github.com/rumart/hpe-g2-pdu-api
        .LINK
            https://www.rudimartinsen.com/2018/11/19/exploring-the-hpe-g2-pdu-rest-api/
        .LINK
            https://www.rudimartinsen.com/2019/01/04/hpe-pdu-powershell-module/
        .EXAMPLE
            PS C:\> Get-HPEPduLoadMeasurement -System 10.10.10.10 -PDUId 1 -Credential $credential


            Current        : 0
            LoadSegmentId  : 1
            BreakerStatus  : Normal
            AppearantPower : 0
            RatedCurrent   : 0
            voltage        : 0
            ActivePower    : 0
            RealPower      : 0
            PowerFactor    : 1
            Energy         : 0

            Current        : 1
            LoadSegmentId  : 2
            BreakerStatus  : Normal
            AppearantPower : 400
            RatedCurrent   : 0
            voltage        : 226
            ActivePower    : 385
            RealPower      : 55016
            PowerFactor    : 0
            Energy         : 0

            Current        : 0
            LoadSegmentId  : 3
            BreakerStatus  : Normal
            AppearantPower : 79
            RatedCurrent   : 0
            voltage        : 226
            ActivePower    : 22
            RealPower      : 38781
            PowerFactor    : 0
            Energy         : 0

            Current        : 0
            LoadSegmentId  : 4
            BreakerStatus  : Normal
            AppearantPower : 0
            RatedCurrent   : 0
            voltage        : 227
            ActivePower    : 0
            RealPower      : 0
            PowerFactor    : 1
            Energy         : 0

            Current        : 0
            LoadSegmentId  : 5
            BreakerStatus  : Normal
            AppearantPower : 0
            RatedCurrent   : 0
            voltage        : 227
            ActivePower    : 0
            RealPower      : 0
            PowerFactor    : 1
            Energy         : 0

            Current        : 0
            LoadSegmentId  : 6
            BreakerStatus  : Normal
            AppearantPower : 0
            RatedCurrent   : 0
            voltage        : 230
            ActivePower    : 0
            RealPower      : 0
            PowerFactor    : 1
            Energy         : 0

    #>
    [cmdletbinding()]
    param(
        [Parameter(Mandatory=$true)]
        $System,
        [Parameter(Mandatory=$true)]
        [int]
        $PDUId,
        [Parameter(Mandatory=$false,ParameterSetName="username")]
        $Username,
        [Parameter(Mandatory=$false,ParameterSetName="username")]
        [securestring]
        $Password,
        [Parameter(Mandatory=$false,ParameterSetName="credential")]
        $Credential,
        [switch]
        $IgnoreSSL
    )
    
    if($Username){
        if(!$Password -or $Password -eq ""){
            $Password = (Read-Host -Prompt "Please specify password" -AsSecureString)
        }
        $Credential = New-Object System.Management.Automation.PSCredential($Username,$Password)
    }
    elseif($Credential){
    }
    else{
        Write-Error "Username or Credential object not specified. Please specify valid credentials for connecting to the PDU"
    }

    $Path = "/redfish/v1/PowerDistribution/$PduId/PowerMeasurement/LoadsegmentMeasurement"

    $response = Invoke-HPEPduRequest -System $System -Resource $Path -Credential $Credential -IgnoreSSL

    if($response.'@odata.type' -eq '#PowerMeasurement.1.0.0.PowerMeasurement'){
        $response.Loadsegments
    }
    
}


HPEPduPS/public/Get-HPEPduOutletMeasurement.ps1

function Get-HPEPduOutletMeasurement {
    <#
        .SYNOPSIS
            Get measurements of HPE Pdu outlets
        .DESCRIPTION
            This function works with HPE G2 Metered and Switched Power distribution units.
            Retrieve information about measurements of the Outlets of the Pdu segment specified.
        .NOTES
            Info
            Author : Rudi Martinsen / Intility AS
            Date : 17/11-2018
            Version : 1.0.0
            Revised : 04/01-2019
            Changelog:
            1.0.0 -- Bumped version number
            0.3.0 -- Setting pduid as a mandatory variable
            0.2.2 -- Added check on credentials
            0.2.1 -- Added help text
            0.2.0 -- Fixed support for credential object
        .LINK
            https://github.com/rumart/hpe-g2-pdu-api
        .LINK
            https://www.rudimartinsen.com/2018/11/19/exploring-the-hpe-g2-pdu-rest-api/
        .LINK
            https://www.rudimartinsen.com/2019/01/04/hpe-pdu-powershell-module/
        .EXAMPLE
            PS C:\> Get-HPEPduOutletMeasurement -System 10.10.10.10 -PDUId 1 -SegmentId 1 -Credential $credential


            AlarmStatus                 : Normal
            OutletNumber                : 1
            EnergyConsumedWattHour      : 860869
            ConsumedCurrent             : 0
            PowerFactor                 : 0
            OutputVoltage               : 226
            LoadPercentage              : 0
            PowerConsumedVoltageAmphere : 156103974
            OutletStatus                : ON
            PowerConsumedWatts          : 148

            AlarmStatus                 : Normal
            OutletNumber                : 2
            EnergyConsumedWattHour      : 140982
            ConsumedCurrent             : 0
            PowerFactor                 : 0
            OutputVoltage               : 226
            LoadPercentage              : 0
            PowerConsumedVoltageAmphere : 39422484
            OutletStatus                : ON
            PowerConsumedWatts          : 18

            AlarmStatus                 : Normal
            OutletNumber                : 3
            EnergyConsumedWattHour      : 94229
            ConsumedCurrent             : 0
            PowerFactor                 : 0
            OutputVoltage               : 226
            LoadPercentage              : 0
            PowerConsumedVoltageAmphere : 223394076
            OutletStatus                : ON
            PowerConsumedWatts          : 218

            AlarmStatus                 : Normal
            OutletNumber                : 4
            EnergyConsumedWattHour      : 666750
            ConsumedCurrent             : 0
            PowerFactor                 : 0
            OutputVoltage               : 226
            LoadPercentage              : 0
            PowerConsumedVoltageAmphere : 40328748
            OutletStatus                : ON
            PowerConsumedWatts          : 3

    #>
    [cmdletbinding()]
    param(
        [Parameter(Mandatory=$true)]
        $System,
        [Parameter(Mandatory=$true)]
        [int]
        $PDUId,
        [Parameter(Mandatory=$true)]
        [int]
        $SegmentId,
        [Parameter(Mandatory=$false,ParameterSetName="username")]
        $Username,
        [Parameter(Mandatory=$false,ParameterSetName="username")]
        [securestring]
        $Password,
        [Parameter(Mandatory=$false,ParameterSetName="credential")]
        $Credential,
        [switch]
        $IgnoreSSL
    )
    
    if($Username){
        if(!$Password -or $Password -eq ""){
            $Password = (Read-Host -Prompt "Please specify password" -AsSecureString)
        }
        $Credential = New-Object System.Management.Automation.PSCredential($Username,$Password)
    }
    elseif($Credential){
    }
    else{
        Write-Error "Username or Credential object not specified. Please specify valid credentials for connecting to the PDU"
    }

    $Path = "/redfish/v1/PowerDistribution/$PduId/PowerMeasurement/Loadsegment/$SegmentId/OutletMeasurement"

    $response = Invoke-HPEPduRequest -System $System -Resource $Path -Credential $Credential -IgnoreSSL

    if($response.'@odata.type' -eq '#OutletMeasurement.1.0.0.OutletMeasurement'){
        $response.Outlets
    }
    
}


HPEPduPS/public/Get-HPEPdu.ps1

function Get-HPEPdu {
    <#
        .SYNOPSIS
            Get information about HPE Pdu(s)
        .DESCRIPTION
            This function works with HPE G2 Metered and Switched Power distribution units.
            Retrieve information about the Pdus connected to the specified Management Ip.
            When pulling all PDUs the output will list the available PDUs. If you specify a PDU Id you will get detailed information about the specified PDU.
        .NOTES
            Info
            Author : Rudi Martinsen / Intility AS
            Date : 17/11-2018
            Version : 1.0.0
            Revised : 04/01-2019
            Changelog:
            1.0.0 -- Bumped version number
            0.2.2 -- Added check on credentials
            0.2.1 -- Added help text
            0.2.0 -- Fixed support for credential object
        .LINK
            https://github.com/rumart/hpe-g2-pdu-api
        .LINK
            https://www.rudimartinsen.com/2018/11/19/exploring-the-hpe-g2-pdu-rest-api/
        .LINK
            https://www.rudimartinsen.com/2019/01/04/hpe-pdu-powershell-module/
        .EXAMPLE
            PS C:\> Get-HPEPdu -System 10.10.10.10 -Username admin

            Id Path                           
            -- ----                           
            1  /redfish/v1/PowerDistribution/1
            2  /redfish/v1/PowerDistribution/2
            3  /redfish/v1/PowerDistribution/3
            4  /redfish/v1/PowerDistribution/4

            Retrieves a list of PDUs

        .EXAMPLE
            PS C:\> Get-HPEPdu -System 10.7.43.115 -PduId 1 -Credential $credential


            Core_location           : 
            @odata.context          : /redfish/v1/$metadata#PowerDistribution
            BreakerRating           : 0
            InputRating             : 16
            Model                   : 230V, 16A, 11.0kVA, 50/60Hz
            LoadsegmentMeasurement  : {@{@odata.id=/redfish/v1/PowerDistribution/1/PowerMeasurement/LoadsegmentMeasurement}}
            Core_u_position         : 
            @odata.type             : #PowerDistribution.1.0.0.PowerDistribution
            Panel_name              : 
            @odata.id               : /redfish/v1/PowerDistribution/1
            OutletMeasurement       : {@{@odata.id=/redfish/v1/PowerDistribution/1/PowerMeasurement/Loadsegment/1/OutletMeasurement}, @{@odata.id=/redfish/v1/PowerDistribution/1/PowerMeasurement/Loadsegment/2/OutletMeasurement}, @{@odata.id=/redfish/v1/PowerDistributio
                                      n/1/PowerMeasurement/Loadsegment/3/OutletMeasurement}, @{@odata.id=/redfish/v1/PowerDistribution/1/PowerMeasurement/Loadsegment/4/OutletMeasurement}...}
            Firmware_version        : 2.0.0.C
            Serial                  : xxxxxx
            Id                      : 1
            PartNumber              : P9S20A
            DeviceType              : PowerDistributionUnit
            PowerDistributionNumber : 1
            Boot_version            : 2.25
            Hardware_version        : HPE
            Voltage                 : 240
            KVARating               : 11
            Power_rating            : 11,0
            OutletControl           : {@{@odata.id=/redfish/v1/PowerDistribution/1/PowerControl/Loadsegment/1/OutletControl}, @{@odata.id=/redfish/v1/PowerDistribution/1/PowerControl/Loadsegment/2/OutletControl}, @{@odata.id=/redfish/v1/PowerDistribution/1/PowerControl
                                      /Loadsegment/3/OutletControl}, @{@odata.id=/redfish/v1/PowerDistribution/1/PowerControl/Loadsegment/4/OutletControl}...}

            
            This will prompt for credentials and connect to the specified IP to output details of the specified PDU
    #>
    [cmdletbinding()]
    param(
        [Parameter(Mandatory=$true)]
        $System,
        [int]
        $PduId,
        [Parameter(Mandatory=$false,ParameterSetName="username")]
        $Username,
        [Parameter(Mandatory=$false,ParameterSetName="username")]
        [securestring]
        $Password,
        [Parameter(Mandatory=$false,ParameterSetName="credential")]
        $Credential,
        [switch]
        $IgnoreSSL = $true
    )

    if($Username){
        if(!$Password -or $Password -eq ""){
            $Password = (Read-Host -Prompt "Please specify password" -AsSecureString)
        }
        $Credential = New-Object System.Management.Automation.PSCredential($Username,$Password)
    }
    elseif($Credential){
    }
    else{
        Write-Error "Username or Credential object not specified. Please specify valid credentials for connecting to the PDU"
    }

    $Path = "/redfish/v1/PowerDistribution"

    if($PDUId){
        $Path += "/$PduId"
    }
    
    $response = Invoke-HPEPduRequest -System $System -Resource $Path -Credential $Credential -IgnoreSSL

    if($response.'@odata.type' -eq '#PowerDistributionCollection.1.0.0.PowerDistributionCollection'){
        $response.members | Select @{l="Id";e={$_."@odata.id".split("/")[-1]}},@{l="Path";e={$_."@odata.id"}}
    }
    elseif($response.'@odata.type' -eq '#PowerDistribution.1.0.0.PowerDistribution'){
        $response
    }
    elseif($response){
        Write-Warning "Unspecified output, the type is $($response.'@odata.type')"
    }
    else{
        #Write-Error "An error occured"
    }

}